api_calls.py

```diff
@@ -366,46 +370,40 @@ def upload_chapter(self, course, book, chapter, args):
                 update_params['video_url_pdf'] = args.video_url
         elif args.slides is not None:
             if not args.slides.lower().endswith(".pdf"):
                 print("Slides argument must be a PDF file.")
                 return False
-            files.append(('input_file_slide', open(self.expand_file_path(args.slides), 'rb')))
+            files.append(('input_file_slide', Path(args.slides)))
             if args.video_url is not None:
                 update_params['video_url_slide'] = args.video_url
         elif args.xml is not None:
             if not (args.xml.lower().endswith(".xml") or args.xml.lower().endswith(".mlx")):
                 print("XML argument must be an XML or MLX file.")
                 return False
-            files.append(('input_file_xml', open(self.expand_file_path(args.xml), 'rb')))
-            if args.images is not None:
-                tmpDir = tempfile.mkdtemp()
-                curPath = os.getcwd()
-                os.chdir(tmpDir)
-                for fg in args.images:
-                    image_path = self.expand_file_path(fg)
-                    if os.path.isdir(image_path):
-                        for root, _, fi in os.walk(image_path):
-                            for name in fi:
-                                full_path = os.path.join(root, name)
-                                shutil.copyfile(full_path, os.path.basename(full_path))
-                    else:
-                        file_glob = glob.glob(image_path)
-                        if len(file_glob) == 0:
-                            print("Invalid input: ", fg)
-                            return False
-                        for f in file_glob:
-                            full_path = self.expand_file_path(f)
-                            shutil.copyfile(full_path, os.path.basename(full_path))
-                for f in os.listdir(os.getcwd()):
-                    print("Uploading file: {}".format(f))
-                    files.append(('image', open(f, 'rb')))
+            files.append(('input_file_xml', Path(args.xml)))
+            if args.attach is not None:
+                for fg in args.attach:
+                    for f in glob.glob(fg):
+                        fg = Path(f)
+                        if fg.is_dir():
+                            # If its a directory, then add all children
+                            files.extend([("image", m) for m in fg.glob("**/*")])
+                        else:
+                            # If its a file, add that directly
+                            files.append(("image", fg))
+
             if args.xml_pdf is not None:
-                files.append(('input_file_xml_pdf', open(self.expand_file_path(args.xml_pdf), 'rb')))
-        response = self.client.post(update_url, headers=headers, files=files, data=update_params)
+                files.append(('input_file_xml_pdf', Path(args.xml_pdf)))
+
+        # Ensure files are closed, even if an error occurs
```

Comment:
I'm not sure we can do this -- the reason I had the whole "copy to a temporary directory and then upload" is because of naming weirdness on diderot.

If you have a file dirA/fileA.png, the file in the XML is referred to as fileA.png, and when diderot is trying to link the correct files in, it expects to find fileA.png inside the request.FILES object. 

If we upload the files like this, then the name of the file sent to diderot is dirA/fileA.png (I believe), and this causes diderot not to find the appropriate files.

However, if this does work, then I'll accept the change!

Reply:
I agree that this behavior makes more sense, but it doesn't agree with the existing Diderot architecture for books -- this has been dictated by umut and anil. This also has to play nicely with how the web uploads work, where the directory based naming wouldn't make the most sense. Taking this change without that renaming step would break uploads for their books I believe. 

I'm not familiar with that library (in general most python + c interop), how does it work?

Aside from this comment I'm happy with the change (and will change the rest of the code to using path and exitstack()).

Reply:
Wait, did I misunderstood at my first pass -- does the path.name return the basename (just the end?) of the path? If so I'm sorry, I'll accept the change then.